qualtran/bloqs/mod_arithmetic/mod_multiplication.py

```diff
@@ -25,54 +26,174 @@
+
+        # Split the three bit pieces again so that we can use the sign to control our constant
+        # addition circuit.
+        x_split = bb.split(x)
+        sign = x_split[0]
+        x = bb.join(x_split[1:], dtype=attrs.evolve(self.dtype, bitsize=self.dtype.bitsize + 1))
+
+        # Add constant p to the x register if the result of the last modular reduction is negative.
+        (sign,), x = bb.add(
+            AddK(bitsize=self.dtype.bitsize + 1, k=self.mod, signed=False, cvs=(1,)),
+            ctrls=(sign,),
+            x=x,
+        )
+
+        # Split the lower bit ancilla from the x register for use in resetting the other ancilla bit
+        # before freeing them both.
+        x_split = bb.split(x)
+        lower_bit = x_split[-1]
+        lower_bit = bb.add(XGate(), q=lower_bit)
+        lower_bit, sign = bb.add(CNOT(), ctrl=lower_bit, target=sign)
+        lower_bit = bb.add(XGate(), q=lower_bit)
+
+        free_bit = x_split[0]
+        x = bb.join(np.concatenate([x_split[1:-1], [lower_bit]]), dtype=self.dtype)
+
+        # Free the ancilla bits.
+        bb.free(free_bit)
+        bb.free(sign)
+
+        # Return the output registers.
+        return {'x': x}
+
+    def wire_symbol(
+        self, reg: Optional['Register'], idx: Tuple[int, ...] = tuple()
+    ) -> 'WireSymbol':
+        if reg is None:
+            return Text(f'x = 2 * x mod {self.mod}')
+        return super().wire_symbol(reg, idx)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {
+            (AddK(self.dtype.bitsize + 2, -self.mod, signed=False), 1),
+            (AddK(self.dtype.bitsize + 1, self.mod, cvs=(1,), signed=False), 1),
+            (CNOT(), 1),
+            (XGate(), 2),
+        }
+
+
+@bloq_example
+def _moddbl_small() -> ModDbl:
+    moddbl_small = ModDbl(QUInt(4), 13)
+    return moddbl_small
+
+
+@bloq_example
+def _moddbl_large() -> ModDbl:
+    prime = 10**9 + 7
+    moddbl_large = ModDbl(QUInt(32), prime)
+    return moddbl_large
+
+
+_MOD_DBL_DOC = BloqDocSpec(bloq_cls=ModDbl, examples=[_moddbl_small, _moddbl_large])
+
+
+@frozen
+class CModMulK(Bloq):
+    r"""Perform controlled modular multiplication by a constant.
+
+    Applies $\ket{c}\ket{c} \rightarrow \ket{c} \ket{x*k^c \mod p}$.
+
+    Args:
+        dtype: Dtype of the register.
         k: The integer multiplicative constant.
         mod: The integer modulus.
-        bitsize: The size of the `x` register.
 
     Registers:
         ctrl: The control bit
         x: The integer being multiplied
     """
 
+    dtype: Union[QUInt, QMontgomeryUInt]
     k: Union[int, sympy.Expr]
     mod: Union[int, sympy.Expr]
-    bitsize: Union[int, sympy.Expr]
 
     def __attrs_post_init__(self):
         if isinstance(self.k, sympy.Expr):
             return
         if isinstance(self.mod, sympy.Expr):
             return
 
-        assert self.k < self.mod
+        assert 0 < self.k < self.mod
```

Comment:
we can't multiply by zero? I think is a legitimate regression re: the modexp test you've xfailed

Reply:
no ... the ModExp thing is wrong ... the only way that ModExp needs to multiply by zero is when the base and mod share a common factor != 1... but if they do then that common factor is a divisor of mod ... so if we are factoring `mod` then we already found a a non trivial factor.

that is the input to ModExp and any factoring subroutine must be numbers coprime to the modulus.

Reply:
and multiplying with zero is just resetting all qubits anyway